File: src/main/java/io/scriptor/http/result/HTTPResultChannel.java
package io.scriptor.http.result;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.channels.ReadableByteChannel;
import java.util.Map;

public class HTTPResultChannel extends HTTPResult<ReadableByteChannel> {

    public HTTPResultChannel(final int statusCode) {
        super(statusCode);
    }

    public HTTPResultChannel(final int statusCode, final @NotNull String statusText) {
        super(statusCode, statusText);
    }

    public HTTPResultChannel(
            final int statusCode,
            final @NotNull String statusText,
            final @NotNull ReadableByteChannel value
    ) {
        super(statusCode, statusText, value);
    }

    public HTTPResultChannel(final int statusCode, final @NotNull Map<String, String> headers) {
        super(statusCode, headers);
    }

    public HTTPResultChannel(
            final int statusCode,
            final @NotNull String statusText,
            final @NotNull Map<String, String> headers
    ) {
        super(statusCode, statusText, headers);
    }

    public HTTPResultChannel(
            final int statusCode,
            final @NotNull String statusText,
            final @NotNull Map<String, String> headers,
            final @NotNull ReadableByteChannel value
    ) {
        super(statusCode, statusText, headers, value);
    }

    @Override
    public int getSize() {
        return -1;
    }

    @Override
    public @NotNull InputStream getStream() {
        return new InputStream() {

            @Override
            public int read(final byte @NotNull [] b, final int off, final int len) throws IOException {
                if (len == 0) {
                    return 0;
                }

                if (getBody().isOpen()) {
                    final var buffer = ByteBuffer.wrap(b).limit(off + len).position(off);
                    return getBody().read(buffer);
                }

                return -1;
            }

            @Override
            public long transferTo(final @NotNull OutputStream out) throws IOException {
                long count = 0;

                while (getBody().isOpen()) {
                    final var buffer = ByteBuffer.wrap(new byte[1024]);
                    final var read   = getBody().read(buffer);

                    if (read < 0) {
                        break;
                    }

                    out.write(buffer.array(), 0, read);
                    count += read;
                }

                return count;
            }

            @Override
            public int read() throws IOException {
                if (getBody().isOpen()) {
                    final var buffer = ByteBuffer.allocate(1);
                    getBody().read(buffer);
                    return buffer.get(0);
                }

                return -1;
            }
        };
    }
}

File: src/main/java/io/scriptor/http/result/HTTPResultChannel.kt
package io.scriptor.http.result

import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.channels.ReadableByteChannel

class HTTPResultChannel : HTTPResult<ReadableByteChannel> {

    override val size = -1
    override val stream = object : InputStream() {

        override fun read(b: ByteArray, off: Int, len: Int): Int {
            if (len == 0) {
                return 0
            }

            if (getBody().isOpen) {
                val buffer = ByteBuffer.wrap(b).limit(off + len).position(off)
                return getBody().read(buffer)
            }

            return -1
        }

        override fun transferTo(out: OutputStream): Long {
            var count: Long = 0

            while (getBody().isOpen) {
                val buffer = ByteBuffer.wrap(ByteArray(1024))
                val read = getBody().read(buffer)

                if (read < 0) {
                    break
                }

                out.write(buffer.array(), 0, read)
                count += read.toLong()
            }

            return count
        }

        override fun read(): Int {
            if (getBody().isOpen) {
                val buffer = ByteBuffer.allocate(1)
                getBody().read(buffer)
                return buffer.get(0).toInt()
            }

            return -1
        }
    }

    constructor(statusCode: Int) : super(statusCode)

    constructor(statusCode: Int, statusText: String) : super(statusCode, statusText)

    constructor(
        statusCode: Int,
        statusText: String,
        value: ReadableByteChannel
    ) : super(statusCode, statusText, value)

    constructor(statusCode: Int, headers: MutableMap<String, String>) : super(statusCode, headers)

    constructor(
        statusCode: Int,
        statusText: String,
        headers: MutableMap<String, String>
    ) : super(statusCode, statusText, headers)

    constructor(
        statusCode: Int,
        statusText: String,
        headers: MutableMap<String, String>,
        value: ReadableByteChannel
    ) : super(statusCode, statusText, headers, value)
}
